lib/segment/src/index/struct_payload_index.rs

```diff
@@ -451,13 +459,28 @@ impl StructPayloadIndex {
                     })
                 })
                 .filter(move |&id| {
-                    !visited_list.check_and_update_visited(id) && struct_filtered_context.check(id)
+                    !visited_list.check_and_update_visited(id) && struct_filtered_check(id)
                 });
 
             Either::Right(iter)
         }
     }
 
+    // TODO make more general by returning a new filter without the primary clauses
+    fn primary_clause_covers_filter<'a>(
+        filter: &'a Filter,
+        query_cardinality: &'a CardinalityEstimation,
+    ) -> bool {
+        let one_condition = filter.total_conditions_count() == 1;
+        let one_primary_clause = query_cardinality.primary_clauses.len() == 1;
+        // the payload index does not know about vector names
+        let no_vector_storage_needed = match query_cardinality.primary_clauses[0] {
+            PrimaryCondition::Condition(_) | PrimaryCondition::Ids(_) => true,
+            PrimaryCondition::HasVector(_) => false,
+        };
```

Comment:
Not sure I understand why those conditions are different. If `HasVector` can be primary (can it?) why we can't use it?

Reply:
I see, but I have concerns that this check introduces second source of truth, which is not immediately obvious how to work with. I am going to try to re-organise it a little, but if it won't help - will merge as-is.

Reply:
For instance, `query_field` function doesn't guarantee that ` PrimaryCondition::Condition` will have an iterator.

Reply:
Good observation about the source of truth :heavy_plus_sign: 

Then we have to call `query_field` first and use the resulting option as the signal for triggering the optimization I believe.

Reply:
I have restructured it in https://github.com/qdrant/qdrant/pull/6789/commits/a605b74b4dad5e135344804205f06875623f9cdd to use properly the results from `query_field`.

As a side effect, it also generates a single full scan iterator when a `query_field` can't do its job instead of potentially one per primary clause.